accounting/invoice/models.py

```diff
@@ -20,3 +20,376 @@
+
+        # Get only the bank account details that exist for the provider.
+        provider_bank_account = self.provider.bank_accounts.first()
+        provider_bank_account_details = tuple(
+            (name, provider_bank_account.identification[key])
+            for key, name in BankAccountIdentifiers.choices
+            if provider_bank_account.identification[key]
+        )
+
+        template = get_template('{template}/{template}.html'.format(template=self.template))
+        invoice = template.render({
+            'site': Site.objects.get_current(),
+            'invoice': self,
+            'users': (
+                ('provider', self.provider),
+                ('client', self.client),
+            ),
+            'bank_account': provider_bank_account,
+            'bank_account_details': provider_bank_account_details,
+            'line_items': aggregated_line_items,
+            'total_quantity': aggregated_quantity['total_quantity'],
+            'total_cost': aggregated_cost['total_cost'],
+            'currency': self.hourly_rate.hourly_rate_currency
+        })
+        pdf_name = 'invoice_{provider}_{client}_{date}.pdf'.format(
+            date=self.date.strftime("%Y-%m-%d"),
+            provider=self.provider.user.username,
+            client=self.client.user.username,
+        )
+        pdf_path = os.path.join(settings.INVOICE_PDF_PATH, pdf_name)
+        pdf_configuration = pdfkit.configuration(wkhtmltopdf=settings.HTML_TO_PDF_BINARY_PATH)
+        pdfkit.from_string(invoice, pdf_path, configuration=pdf_configuration, options=self.PDF_OPTIONS)
+        return pdf_path
+
+    def upload_to_google_drive(self, invoice_file):  # NOQA
+        """
+        Upload the invoice in some file format to Google Drive.
+
+        Assumes the invoice has already been turned into a file, i.e. a PDF.
+        Assumes a _very_ specific file structure right now. For example:
+            |- 2016
+            |- 2017
+            |   |- blah
+            |   |- ...
+            |   |- invoices-in
+            |   |   |- 1
+            |   |   |- 2
+            |   |   |- ...
+            |   |   |- 12
+            |   |   |   |- invoice_1.pdf
+            |   |   |   |- ...
+            |   |   |   |- invoice_n.pdf
+            |- 2018
+
+        TODO: This should be a synchronous job on a non-web worker.
+        """
+        if not settings.ENABLE_GOOGLE:
+            return
+
+        # Generate a query we'll use repeatedly for finding the proper folder.
+        query = ("mimeType ='application/vnd.google-apps.folder' and "
+                 "'{}' in parents and "
+                 "trashed=false")
+
+        def get_folder_id(root, name):
+            """ Returns the ID of the folder with name `name` in the folder given by ID `root`. """
+            folders = drive.ListFile({'q': query.format(root)}).GetList()
+            for folder in folders:
+                if folder['title'] == name:
+                    return folder['id']
```

Comment:
@UmanShahzad It shouldn't be too hard to make this more flexible. Here's a sketch for an implementation that uses recursion to support arbitrary paths:

```python
def get_folder_id(root, path=[]):
    if path:
        name = path[0]
        folders = drive.ListFile({'q': query.format(parent=root)}).GetList()
        for folder in folders:
            if folder['title'] == name:
                root = folder['id']
                break
        return get_folder_id(root, path[1:])
    else:
        return root
```

Below we could then just do:

```python
target_folder = get_folder_id(
    settings.GOOGLE_DRIVE_ROOT,
    path=[str(self.billing_start_date.year), 'invoices-in', str(self.billing_start_date.month)]
)
```

Or even:

```python
target_folder = get_folder_id(settings.GOOGLE_DRIVE_ROOT, path=self.google_drive_path)
```

... if we add a `google_drive_path` property that returns `[str(self.billing_start_date.year), 'invoices-in', str(self.billing_start_date.month)]`.

Reply:
Ah, nice! Yes I'll definitely use this :+1: